sematic/api/endpoints/artifacts.py

```diff
@@ -70,9 +72,55 @@ def get_artifact_endpoint(user: Optional[User], artifact_id: str) -> flask.Respo
             "No Artifact with id {}".format(repr(artifact_id)), HTTPStatus.NOT_FOUND
         )
 
-    payload = dict(
-        content=artifact.to_json_encodable(),
-    )
+    payload = dict(content=artifact.to_json_encodable())
+
+    return flask.jsonify(payload)
+
+
+@sematic_api.route("/api/v1/artifacts/cache", methods=["GET"])
+@authenticate
+def get_cached_artifact_endpoint(user: Optional[User]) -> flask.Response:
+    """
+    Retrieve an Artifact from the cache, along with the Run that originally produced it.
+
+    Request
+    -------
+    cache_key: str
+        The cache key under which to look for the Artifact.
+
+    Response
+    --------
+    artifact: Artifact
+        The found Artifact in JSON format.
+    run: Run
+        The Run that originally produced the Artifact, in JSON format.
+    """
+    if CACHE_KEY not in set(flask.request.args.keys()):
+        return jsonify_error(
+            f"Can only get artifacts using the query key {CACHE_KEY}",
+            HTTPStatus.BAD_REQUEST,
+        )
+
+    if len(flask.request.args.keys()) > 1:
+        unknown_keys = set(flask.request.args.keys()) - {CACHE_KEY}
+        return jsonify_error(
+            f"Unknown query keys: {unknown_keys}", HTTPStatus.BAD_REQUEST
+        )
+
+    cache_key = flask.request.args[CACHE_KEY]
+    payload: Dict[str, Optional[Dict[str, str]]] = dict(content=None)
+
+    artifact_and_run = get_cached_artifact_and_run(cache_key)
+
+    if artifact_and_run is None:
+        # not finding an artifact is part of the normal flow and not an error case
+        return flask.jsonify(payload)
+
+    artifact, run = artifact_and_run
+    payload["content"] = {
+        "artifact": artifact.to_json_encodable(),
+        "run": run.to_json_encodable(),
+    }
```

Comment:
Instead of this endpoint, can we instead obtain the original run ID, then query its graph to get the output artifact?
Basically:
- query `/api/v1/runs?filters=[["cache_key", "eq", <cache_key>]]&limit=1`, with maybe a sorting parameter to make sure we get the earliest run with this cache key
- then `/api/v1/runs/<run_id>/graph?root=False`

Why two queries instead of one?
Ideally the API server is very generic and maximally agnostic to particular features. It is a CRUD resource server. If we create new endpoints for particular queries as opposed to using combinations of existing endpoints (as long as it is reasonable performance-wise), we will balloon the number of "special" "single-feature" endpoints, and it is very bad for code complexity.

So what are we trying to achieve here? We have a run that is cached, and we know it's cache key. So first, we want to find the original run. We do this by querying the run search endpoint with the appropriate query parameters.
Once we have found the original run, we ask for its output artifact. The way to do this currently is to query the graph endpoint which by default `root=False` will only return the "local" graph, i.e. the run, it's inputs and output artifacts.

This also mean we don't need new functions in `queries` or `api_client`.

Performance-wise, this is acceptable. The first query yields a single `SELECT` and returns a single run. The second query does one `SELECT` for the run, then one (maybe two?) for the artifacts. This is reasonable and reduces code complexity.

Reply:
Done - this feature is implemented in #470, and this PR rebased on top if that one.